### PR TITLE
[FEATURE] Autocomplete for AppInstances + more tests

### PR DESCRIPTION
**Pull request by Kanti**
resolves: #49

### CODE REVIEW

File: src/Cli/PhpsuApplication.php [hunk 1]
```diff
@@ -3,16 +3,23 @@
...
 final class PhpsuApplication
 {
     public static function command(): void
     {
-        $application = new Application('phpsu', '1.0.0-alpha2');
-        $command = new SyncCliCommand();
-        $application->add($command);
-        $application->add(new SshCliCommand());
-        $application->run();
+        self::createApplication()->run();
```
Comment (ChrisB9) [inline src/Cli/PhpsuApplication.php]: why did you move that?

File: src/Command/DatabaseCommand.php [hunk 2]
```diff
@@ -50,6 +50,12 @@ public static function fromGlobal(GlobalConfig $global, string $fromInstanceName
             }
             $result[] = static::fromAppInstances($fromInstance, $toInstance, $fromDatabase, $toDatabase, $currentHost, $all);
         }
+        foreach ($fromInstance->getDatabases() as $databaseName => $fromDatabase) {
+            if ($toInstance->hasDatabase($databaseName)) {
```
Comment (ChrisB9) [inline src/Command/DatabaseCommand.php]: i stumbled upon this while reading: i know toInstance is meant as its the instance where things are going to but while reading it, "toInstance" sounds like you made / make something to an instance.. I'm aware this is not really an issue but i just wanted to point this out

File: src/Controller.php [hunk 3]
```diff
@@ -9,53 +9,39 @@
...
-     * @var OutputInterface
-     */
-    private $output;
-    /**
-     * @var GlobalConfig
-     */
-    private $config;
 
-    public function __construct(OutputInterface $output, GlobalConfig $config)
+    public function ssh(OutputInterface $output, GlobalConfig $config, string $destination, string $currentHost, string $command, bool $dryRun): int
```
Comment (ChrisB9) [inline src/Controller.php]: why do want to make OutputInterface and GlobalConfig as a parameter for every method?

File: src/ControllerInterface.php [hunk 4]
```diff
@@ -0,0 +1,14 @@
...
+namespace PHPSu;
+
+use PHPSu\Config\GlobalConfig;
+use Symfony\Component\Console\Output\OutputInterface;
+
+interface ControllerInterface
+{
+    public function ssh(OutputInterface $output, GlobalConfig $config, string $destination, string $currentHost, string $command, bool $dryRun): int;
+
+    public function sync(OutputInterface $output, GlobalConfig $config, string $form, string $to, string $currentHost, bool $dryRun, bool $all, bool $noFiles, bool $noDatabases): void;
```
Comment (ChrisB9) [inline src/ControllerInterface.php]: it is going to be ridiculous adding more parameters like that :D

File: src/Helper/StringHelper.php [hunk 5]
```diff
@@ -26,4 +26,15 @@ public static function splitString(string $string, int $maxLength): array
         }
         return [$full];
     }
+
+    public static function findStringInArray(string $needle, array $haystack): ?string
```
Comment (ChrisB9) [inline src/Helper/StringHelper.php]: isnt that kind of, what http://php.net/manual/de/function.array-search.php is for?

File: tests/Cli/PhpsuApplicationTest.php [hunk 6]
```diff
@@ -0,0 +1,19 @@
...
+use PHPSu\Cli\PhpsuApplication;
+use PHPSu\Cli\SshCliCommand;
+use PHPSu\Cli\SyncCliCommand;
+use PHPUnit\Framework\TestCase;
+
+class PhpsuApplicationTest extends TestCase
+{
+    public function testCommand()
+    {
+        $app = PhpsuApplication::createApplication();
```
Comment (ChrisB9) [inline tests/Cli/PhpsuApplicationTest.php]: so, command basically is going to be useless in the future?

File: tests/Config/SshUrlTest.php [hunk 7]
```diff
@@ -8,6 +8,28 @@
...
+    }
+
+    public function testInvalidUser(): void
+    {
+        $this->expectException(\Exception::class);
+        $this->expectExceptionMessageRegExp('/User must be set/');
+        new SshUrl('test');
+    }
+
+    public function testInvalidPor(): void
```
Comment (ChrisB9) [inline tests/Config/SshUrlTest.php]:
```suggestion
    public function testInvalidPort(): void
```

File: tests/ControllerTest.php [hunk 8]
```diff
@@ -28,8 +27,183 @@ public function testEmptyConfigSyncDryRun(): void
...
+
+        $output = new BufferedOutput();
+        $controller = new Controller();
+        $controller->sync($output, $config, 'testing', 'local', '', true, true, true, false);
+        $lines = [
+            'database:database',
+            "ssh -F '.phpsu/config/ssh_config' 'projectEu' 'mysqldump --opt --skip-comments -h'\''127.0.0.1'\'' -u'\''test'\'' -p'\''aaaaaaaa'\'' '\''testdb'\''' | mysql -h'127.0.0.1' -u'root' -p'root' 'test1234'",
+            '',
+        ];
+        $this->assertSame($lines, explode("\n", $output->fetch()));
```
Comment (ChrisB9) [inline tests/ControllerTest.php]:
```suggestion
        $this->assertSame($lines, explode(PHP_EOL, $output->fetch()));
```

File: src/Helper/StringHelper.php (hunk 5 above)
Comment (Kanti) [reply src/Helper/StringHelper.php]: no it is not the same function

File: tests/Cli/PhpsuApplicationTest.php (hunk 6 above)
Comment (Kanti) [reply tests/Cli/PhpsuApplicationTest.php]: why?

File: src/Cli/PhpsuApplication.php (hunk 1 above)
Comment (Kanti) [reply src/Cli/PhpsuApplication.php]: for code coverage

File: src/Command/DatabaseCommand.php (hunk 2 above)
Comment (Kanti) [reply src/Command/DatabaseCommand.php]: ``InstanceTo``? is worse

Comment (ChrisB9) [reply src/Command/DatabaseCommand.php]: alternatively: source, destination - but ofc thats longer

File: tests/Cli/PhpsuApplicationTest.php (hunk 6 above)
Comment (ChrisB9) [reply tests/Cli/PhpsuApplicationTest.php]: because it currently is calling just another method

Comment (Kanti) [reply tests/Cli/PhpsuApplicationTest.php]: it calls 2 methods
